webhook/mutation/pod_mutator.go

```diff
@@ -204,264 +224,474 @@ func isWellKnownWorkload(ownerRef metav1.OwnerReference) bool {
-			return admission.Errored(http.StatusBadRequest, err)
-		}
-	} else if err != nil {
-		log.Error(err, "failed to query the init secret before pod injection")
-		return admission.Errored(http.StatusBadRequest, err)
+	secretResponse := m.ensureInitSecret(ctx, ns, dk)
+	if secretResponse != nil {
+		return *secretResponse
 	}
 
-	endpointGenerator := dtingestendpoint.NewEndpointGenerator(m.client, m.apiReader, m.namespace, log)
-
-	var endpointSecret corev1.Secret
-	if err := m.apiReader.Get(ctx, client.ObjectKey{Name: dtingestendpoint.SecretEndpointName, Namespace: ns.Name}, &endpointSecret); k8serrors.IsNotFound(err) {
-		if _, err := endpointGenerator.GenerateForNamespace(ctx, dkName, ns.Name); err != nil {
-			log.Error(err, "failed to create the data-ingest endpoint secret before pod injection")
-			return admission.Errored(http.StatusBadRequest, err)
-		}
-	} else if err != nil {
-		log.Error(err, "failed to query the data-ingest endpoint secret before pod injection")
-		return admission.Errored(http.StatusBadRequest, err)
+	dataIngestFields, diResponse := m.ensureDataIngestSecret(ctx, ns, dkName, dk)
+	if diResponse != nil {
+		return *diResponse
 	}
 
-	dataIngestFields, err := endpointGenerator.PrepareFields(ctx, &dk)
-	if err != nil {
-		log.Error(err, "failed to query the data-ingest endpoint secret before pod injection")
-		return admission.Errored(http.StatusBadRequest, err)
+	log.Info("injecting into Pod", "name", pod.Name, "generatedName", pod.GenerateName, "namespace", req.Namespace)
+
+	response := m.handleAlreadyInjectedPod(pod, dk, injectionInfo, dataIngestFields, req)
+	if response != nil {
+		return *response
 	}
 
-	log.Info("injecting into Pod", "name", pod.Name, "generatedName", pod.GenerateName, "namespace", req.Namespace)
+	injectionInfo.fillAnnotations(pod)
 
-	if pod.Annotations == nil {
-		pod.Annotations = map[string]string{}
+	workloadName, workloadKind, workloadResponse := m.retrieveWorkload(ctx, req, injectionInfo, pod)
+	if workloadResponse != nil {
+		return *workloadResponse
 	}
 
-	if pod.Annotations[dtwebhook.AnnotationInjected] == "true" {
-		if dk.FeatureEnableWebhookReinvocationPolicy() {
-			var needsUpdate = false
-			var installContainer *corev1.Container
-			for i := range pod.Spec.Containers {
-				c := &pod.Spec.Containers[i]
-
-				preloaded := false
-				for _, e := range c.Env {
-					if e.Name == "LD_PRELOAD" {
-						preloaded = true
-						break
-					}
-				}
+	technologies, installPath, installerURL, failurePolicy, image := m.getBasicData(pod)
 
-				if !preloaded {
-					// container does not have LD_PRELOAD set
-					log.Info("instrumenting missing container", "name", c.Name)
+	dkVol, mode := ensureDkVolume(dk)
 
-					deploymentMetadata := deploymentmetadata.NewDeploymentMetadata(m.clusterID, deploymentmetadata.DeploymentTypeApplicationMonitoring)
-					updateContainer(c, &dk, pod, deploymentMetadata, dataIngestFields)
+	ensureInjectionConfigVolume(pod)
```

Comment:
`ensure` sounds like the function verifies the volumes exist. would `setup` work as well?